### ISSUE TITLE
Tile Join seems incredibly slow when merging tilesets

### ISSUE THREAD

**Issue by dylanparry** [mapbox/tippecanoe#448]
Hi,

I'm attempting to merge approximately 20 mbtiles files to produce one tileset. This works fine, and results in a 4.5GB mbtiles file, but is very slow to produce.

The behaviour I am seeing is that for the first ~200MB there is very little disk activity and equally little CPU usage. There is then a sudden spike to ~20MB/s disk writing and full processor usage up to about 1.8GB of file being written. It then returns to low disk activity and CPU for some time before it spikes again and completes most of the file.

It seems to me that a large portion of this "inactive" time is spent working through tiles that have no data, ie. those that are outside of the areas covered by the mbtiles files. I have created coverage only for the UK, and so most of the world has no data at all. During this time, the resultant mbtiles file doesn't appear to grow in size (or very little at least).

I recall a similar issue with writing the mbtiles files in the first place using an older version of Tippecanoe, but that seems to have been fixed, presumably by checking the extents of the GeoJSON and only attempting to generate tiles for areas that fall inside those extents? Are any similar tests being done when it comes to joining tiles, as it would appear that this is not the case.

Thanks,

Dylan

**Comment by ericfischer** [comment]
Thanks for the report. Tile-join only reads the tiles that are in the source mbtiles, so it should not spend any time on areas that don't have any tiles.

The up-front thing that could be expensive is that it does

     SELECT zoom_level, tile_column, tile_row, tile_data from tiles order by zoom_level, tile_column, tile_row;

on each mbtiles in order to get the tiles in the same order from each, so there might be some time spent on sorting, but I would expect that to manifest as CPU (or possibly disk access) time rather than nothing happening at all.

It tries to read 100 times as many tiles into memory as there are CPUs before starting the merge, so that could be a bottleneck if it means something else in memory has to get paged out to make room for those tiles, or 100 might be an unnecessarily large queue. (It has to queue up at least some number of tasks, or there is more overhead from creating threads than there is from parallelism.)

How many CPUs are on the computer you are using?

**Comment by dylanparry** [comment]
I have a quad core i7 processor, so eight threads altogether. There's 32GB of RAM on this laptop too, and it didn't look like it was using more than 4GB in total, and the swap partition was barely being used (60MB maybe).

**Comment by dylanparry** [comment]
I think this might be a Docker issue. I have just compiled and installed Tippecanoe directly on my laptop (Ubuntu Gnome 17.04) and the same data took just under ten minutes to process, whereas running it in Docker took at least 2 hours (I got tired and left it running).

Any idea why this might be happening? I'm happy to continue running Tippecanoe compiled from source now that I've already done it, but this might affect future users who would prefer to stick with Docker.
